src/org/labkey/test/components/glassLibrary/grids/ResponsiveGrid.java

```diff
@@ -186,100 +194,52 @@ public ResponsiveGrid selectAllOnPage(boolean checked)
+    public List<GridRow> getSelectedRows()
     {
-        return getRows().get(index);
+        return new GridRow.GridRowFinder(this).withCheckedBox().findAll(this);
     }
 
-    protected WebElement getCell(int rowIndex, int colIndex)
+    private GridRow getRow(int index)
     {
-        return Locator.tag("td").index(colIndex).findElement(getRow(rowIndex));
+        return new GridRow.GridRowFinder(this).index(index).find(this);
     }
 
-    public Integer getRowIndex(String containsText)
+    public Optional<GridRow> getRow(String containsText)
     {
-        List<WebElement> rows = getRows();
-        for(int i=0; i<rows.size(); i++ )
-        {
-            if (getRow(i).getText().contains(containsText))
-            {
-                return i;
-            }
-        }
-        return -1;
+        return new GridRow.GridRowFinder(this).withValue(containsText).findOptional(this);
     }
 
-    public Integer getRowIndex(String containsText, String searchColumnHeader)
+    public Optional<GridRow> getRow(String containsText, String columnHeader)
     {
-        List<WebElement> rows = getRows();
-        int colIndex = getColumnIndex(searchColumnHeader);
-        for(int i=0; i<rows.size(); i++ )
-        {
-            // find the nth td in the row, see if its text matches
-            if (getCell(i, colIndex).getText().equalsIgnoreCase(containsText))
-            {
-                return i;
-            }
-        }
-        return -1;
+        // try to normalize column index to start at 0, excluding row selector column
+        Integer columnIndex = getColumnIndex(columnHeader);
+        return new GridRow.GridRowFinder(this).withValueAtColumnIndex(containsText, columnIndex)
+                .findOptional(this);
     }
 
-    public List<String> getColumnDataAsText(String columnHeader)
+    public Optional<GridRow> getRow(Locator.XPathLocator containing)
     {
-        List<String> columnData = new ArrayList<>();
-        Integer columnIndex = getColumnIndex(columnHeader);
-
-        for (int i = 0; i < getRows().size(); i++)
-        {
-            columnData.add(getCell(i, columnIndex).getText());
-        }
-
-        return columnData;
+        return new GridRow.GridRowFinder(this).withDescendant(containing).findOptional();
     }
 
-    public WebElement getRow(String exactColumnText, String columnHeader)
+    public List<GridRow> getRows()
     {
-        int colIndex = getColumnIndex(columnHeader);
-        return Locator.xpath("//tbody/tr[./td[" + (colIndex + 1) + "]/span[text()=normalize-space('" + exactColumnText + "')]]")
-                .findOptionalElement(this).orElseThrow(() -> new NoSuchElementException("Could not find a row where column '" + columnHeader + "' had a value of '" + exactColumnText + "'."));
+        return new GridRow.GridRowFinder(this).findAll(getComponentElement());
     }
 
-    public Integer getRowIndex(Locator containing)
+    protected WebElement getCell(int rowIndex, String columnHeader)
```

Comment:
Kind of a big change from getRowIndex to getCell.

Reply:
Yes, it was a big change to eliminate getRowIndex- fortunately, there weren't many usages of it and I refactored them out to have tests rely on object references rather than having to know how indexes work in the row.

Also, by the time I was done refactoring, there were also no usages of getCell